backend/src/hooks/check-has-permission/check-has-permission.hook.ts

```diff
@@ -0,0 +1,41 @@
+import { FastifyRequest } from 'fastify';
+
+import {
+  ExceptionMessage,
+  HttpCode,
+  PermissionKey,
+} from '~/common/enums/enums';
+import { EntityPaginationRequestQueryDto } from '~/common/types/types';
+import { PermissionsError } from '~/exceptions/exceptions';
+import { checkHasPermission } from '~/helpers/helpers';
+
+const checkHasPermissions =
+  (...pagePermissions: PermissionKey[]) =>
+  async (
+    req: FastifyRequest<{ Querystring: EntityPaginationRequestQueryDto }>,
+  ): Promise<void> => {
+    const { user } = req;
+    const hasUser = Boolean(user);
+    const userPermissions = user.permissions.map((item) => item.key);
```

Comment:
If there's no user, `user.permissions` will throw an error, I think it's better to move it to after `if (hasUser) { throw ... }`